File: indiaquant/packages/finance-math/src/xirr.ts
import Decimal from "decimal.js";

/**
 * XIRR — Extended Internal Rate of Return (Money-Weighted Return).
 *
 * Uses Newton-Raphson iteration to find the rate r such that:
 *   NPV = Σ [ cashflow_i / (1 + r)^(days_i / 365) ] = 0
 *
 * All monetary values use Decimal to avoid floating-point drift.
 */

export interface Cashflow {
  /** ISO date string YYYY-MM-DD */
  date: string;
  /** Negative = outflow (investment), Positive = inflow (redemption/current value) */
  amount: number;
}

const MAX_ITERATIONS = 1000;
const TOLERANCE = 1e-10;

/**
 * Calculate XIRR for a series of cashflows.
 * Returns null if no solution is found (e.g., all cashflows same sign).
 */
export function calculateXIRR(cashflows: Cashflow[]): number | null {
  if (cashflows.length < 2) return null;

  const dates = cashflows.map((cf) => new Date(cf.date).getTime());
  const amounts = cashflows.map((cf) => cf.amount);
  const firstDate = dates[0];

  // Days from first cashflow for each cashflow
  const dayOffsets = dates.map((d) => (d - firstDate) / (1000 * 60 * 60 * 24));

  /**
   * NPV function: Σ [ amount_i / (1 + rate)^(days_i / 365) ]
   */
  function npv(rate: number): number {
    return amounts.reduce((sum, amount, i) => {
      const exponent = dayOffsets[i] / 365;
      return sum + amount / Math.pow(1 + rate, exponent);
    }, 0);
  }

  /**
   * Derivative of NPV with respect to rate (for Newton-Raphson).
   */
  function npvDerivative(rate: number): number {
    return amounts.reduce((sum, amount, i) => {
      const t = dayOffsets[i] / 365;
      return sum - (t * amount) / Math.pow(1 + rate, t + 1);
    }, 0);
  }

  // Initial guess: 10%
  let rate = 0.1;

  for (let i = 0; i < MAX_ITERATIONS; i++) {
    const npvValue = npv(rate);
    const derivative = npvDerivative(rate);

    if (Math.abs(derivative) < 1e-15) return null; // Avoid division by zero

    const newRate = rate - npvValue / derivative;

    if (Math.abs(newRate - rate) < TOLERANCE) {
      return newRate;
    }

    rate = newRate;

    // Guard against divergence
    if (rate < -0.999 || rate > 100) return null;
  }

  return null; // Did not converge
}

/**
 * Calculate absolute return: (current_value - total_invested) / total_invested
 */
export function calculateAbsoluteReturn(
  totalInvested: string,
  currentValue: string
): number {
  const invested = new Decimal(totalInvested);
  const current = new Decimal(currentValue);
  if (invested.isZero()) return 0;
  return current.minus(invested).div(invested).toNumber();
}

/**
 * Calculate CAGR: (end_value / start_value)^(1 / years) - 1
 */
export function calculateCAGR(
  startValue: number,
  endValue: number,
  years: number
): number | null {
  if (startValue <= 0 || years <= 0) return null;
  return Math.pow(endValue / startValue, 1 / years) - 1;
}

/**
 * Calculate Time-Weighted Return (TWR) from a series of sub-period returns.
 * Each sub-period return is (end_value / start_value) - 1.
 */
export function calculateTWR(subPeriodReturns: number[]): number {
  if (subPeriodReturns.length === 0) return 0;
  const product = subPeriodReturns.reduce((acc, r) => acc * (1 + r), 1);
  return product - 1;
}

File: indiaquant/apps/analytics-engine/routers/xirr.py
"""
XIRR / Performance router.
Newton-Raphson XIRR, TWR, and contribution analysis.
"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
import numpy as np

router = APIRouter()


class CashflowItem(BaseModel):
    date: str  # YYYY-MM-DD
    amount: float  # negative = outflow, positive = inflow


class XIRRRequest(BaseModel):
    cashflows: list[CashflowItem]


class XIRRResponse(BaseModel):
    xirr: float | None
    twr: float | None
    absolute_return: float
    cagr: float | None


@router.post("/calculate", response_model=XIRRResponse)
async def calculate_xirr(request: XIRRRequest):
    """Calculate XIRR, TWR, absolute return, and CAGR from cashflows."""
    try:
        if len(request.cashflows) < 2:
            raise HTTPException(status_code=400, detail="At least 2 cashflows required")

        from datetime import datetime

        dates = [datetime.strptime(cf.date, "%Y-%m-%d") for cf in request.cashflows]
        amounts = [cf.amount for cf in request.cashflows]
        first_date = dates[0]
        day_offsets = [(d - first_date).days for d in dates]

        def npv(rate: float) -> float:
            return sum(
                amt / (1 + rate) ** (days / 365)
                for amt, days in zip(amounts, day_offsets)
            )

        def npv_derivative(rate: float) -> float:
            return sum(
                -(days / 365) * amt / (1 + rate) ** (days / 365 + 1)
                for amt, days in zip(amounts, day_offsets)
            )

        # Newton-Raphson
        rate = 0.1
        xirr = None
        for _ in range(1000):
            npv_val = npv(rate)
            deriv = npv_derivative(rate)
            if abs(deriv) < 1e-15:
                break
            new_rate = rate - npv_val / deriv
            if abs(new_rate - rate) < 1e-10:
                xirr = new_rate
                break
            rate = new_rate
            if rate < -0.999 or rate > 100:
                break

        # Absolute return
        outflows = sum(abs(a) for a in amounts if a < 0)
        inflows = sum(a for a in amounts if a > 0)
        absolute_return = (inflows - outflows) / outflows if outflows > 0 else 0

        # CAGR
        years = (dates[-1] - dates[0]).days / 365
        cagr = None
        if outflows > 0 and inflows > 0 and years > 0:
            cagr = (inflows / outflows) ** (1 / years) - 1

        return XIRRResponse(
            xirr=round(xirr, 6) if xirr is not None else None,
            twr=None,  # TWR requires sub-period returns — computed client-side
            absolute_return=round(absolute_return, 6),
            cagr=round(cagr, 6) if cagr is not None else None,
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
